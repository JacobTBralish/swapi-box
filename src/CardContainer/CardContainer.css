.card-container {
  width: 90%;
  height: 100%;
  display: flex;
  flex-flow: wrap row;
  align-items: center;
  justify-content: center;
  margin: 10px auto;
}