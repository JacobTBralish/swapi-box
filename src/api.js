const url = 'https://swapi.co/api';

export const fetchApi = async (url) => {
  const fetched = await fetch(url)
  const jsonFetch = await fetched.json()
    
  return jsonFetch;
}

export const getFilmNum = () => {
  return Math.floor(Math.random() * 7 + 1)
}

export const getScroll = async () => {
  const random = getFilmNum()
  const {title, opening_crawl, release_date} = await fetchApi(`${url}/films/${random}`)
   
  return { title, opening_crawl, release_date }
}

export const getPeople = async () => {
  const { results } = await fetchApi(`${url}/people`)
  const completePeople = getMorePeopleData(results)

  return completePeople
}

const getMorePeopleData = (people) => {
  const completePeople = people.map( async ({ name, homeworld, species }) => {
    homeworld = await fetchApi(homeworld)
    species = await fetchApi(species[0])
    return { 
      Name: name , 
      Homeworld: homeworld.name,
      Population: homeworld.population, 
      Species: species.name}
  })

  return Promise.all(completePeople)
}

export const getPlanets = async () => {
  const { results } = await fetchApi(`${url}/planets`);
  const completePlanets = getMorePlanetData(results);

  return completePlanets;
}

const getMorePlanetData = (planets) => {
  const completePlanets = planets.map( async ({ name, residents, terrain, population}) => {
    const retrievedResidents = await getResidents(residents)

    return {
      Name: name, 
      Residents: retrievedResidents, 
      Terrain: terrain, 
      Population: population
    }
  })

  return Promise.all(completePlanets)
}

const getResidents = (residents) => {
  const retrievedResidents = residents.map( async resident => {
    const { name } = await fetchApi(resident)
    return name; 
  })

  return Promise.all(retrievedResidents)
}

export const getVehicles = async () => {
  const { results } = await fetchApi(`${url}/vehicles`)
  const vehicleArray = results.map( ({name, model, vehicle_class, passengers}) => {
    return {
      Name: name, 
      Model: model, 
      Class : vehicle_class, 
      Passengers: passengers
    }
  })

  return vehicleArray;
}
