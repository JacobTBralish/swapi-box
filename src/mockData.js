export const mockPeople = [
    {
      "name": "Luke Skywalker",
      "height": "172",
      "mass": "77",
      "hair_color": "blond",
      "skin_color": "fair",
      "eye_color": "blue",
      "birth_year": "19BBY",
      "gender": "male",
      "homeworld": "https://swapi.co/api/planets/1/",
      "films": [
        "https://swapi.co/api/films/2/",
        "https://swapi.co/api/films/6/",
        "https://swapi.co/api/films/3/",
        "https://swapi.co/api/films/1/",
        "https://swapi.co/api/films/7/"
      ],
      "species": [
        "https://swapi.co/api/species/1/"
      ],
      "vehicles": [
        "https://swapi.co/api/vehicles/14/",
        "https://swapi.co/api/vehicles/30/"
      ],
      "starships": [
        "https://swapi.co/api/starships/12/",
        "https://swapi.co/api/starships/22/"
      ],
      "created": "2014-12-09T13:50:51.644000Z",
      "edited": "2014-12-20T21:17:56.891000Z",
      "url": "https://swapi.co/api/people/1/"
    }
  ]

  export const mockFilm = {
  "title": "A New Hope",
  "episode_id": 4,
  "opening_crawl": "It is a period of civil war.\r\nRebel spaceships, striking\r\nfrom a hidden base, have won\r\ntheir first victory against\r\nthe evil Galactic Empire.\r\n\r\nDuring the battle, Rebel\r\nspies managed to steal secret\r\nplans to the Empire's\r\nultimate weapon, the DEATH\r\nSTAR, an armored space\r\nstation with enough power\r\nto destroy an entire planet.\r\n\r\nPursued by the Empire's\r\nsinister agents, Princess\r\nLeia races home aboard her\r\nstarship, custodian of the\r\nstolen plans that can save her\r\npeople and restore\r\nfreedom to the galaxy....",
  "release_date": "1977-05-25"
}