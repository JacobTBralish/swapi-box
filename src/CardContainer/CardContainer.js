import React, { Component } from 'react'
import { Card } from '../Card/Card'
import './CardContainer.css'


class CardContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      people: JSON.parse(localStorage.getItem('people')) || [],
      planets: JSON.parse(localStorage.getItem('planets')) || [],
      vehicles: JSON.parse(localStorage.getItem('vehicles')) || [],
      favorites: [],
      loading: null
    }
  }

  addFavoriteProp = (card) => {
    return {...card, style: 'favorite'}
  }

  findSelected = (event, card) => {
    return event.target.id === card.Name
  }

  handleClick = (event) => {
    const { name } = this.props

    if(!event.target.classList.contains('favorite')) {
    
      const cardToFavorite = this.state[name].find( card => this.findSelected(event, card))
      const notFavorites = this.state[name].filter( card => event.target.id !== card.Name )

      const newCard = this.addFavoriteProp(cardToFavorite)

      const newState = [...notFavorites, newCard]

      const currentFavorites = [...this.state.favorites]

      this.setState({
        [name] : newState,
        favorites: [...currentFavorites, newCard]
      }, localStorage.setItem([name], JSON.stringify(newState)))
    }
  }

  async componentDidMount() {
    const { name, fetch } = this.props

    if(!this.state[name].length) {
      try {
        this.setState({ loading: true })
        const newState = await fetch();

        this.setState({
          [name] : newState,
          loading: false
        }, localStorage.setItem([name], JSON.stringify(newState)))
      } catch(error) {
          throw(new Error('Error retrieving sick Star Wars Data'))
      }
    }
  }

  async componentWillReceiveProps(nextProps) {
    const { name, fetch } = nextProps

    if(this.props.name !== name && !this.state[name].length) {
      try {
        this.setState({ loading: true })
        const newState = await fetch()

        this.setState({
          [name] : newState,
          loading: false
        }, localStorage.setItem([name], JSON.stringify(newState)))
      } catch(error) {
        throw(new Error('Error retrieving sick Star Wars Data'))
      }
    }
  }

  render() {
    const { name } = this.props

    const Cards = this.state[name].map( (card, i) => {
      return <Card {...card} 
              key={card.Name + i} 
              handleClick={this.handleClick}
              id={card.Name} /> 
    })

    return (
      <div className='card-container'>
        { this.state.loading && <h1> Loading... </h1> }
        {Cards}
      </div>
    )
  }
}

export default CardContainer