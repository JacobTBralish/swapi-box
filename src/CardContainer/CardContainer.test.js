import React from 'react';
import ReactDOM from 'react-dom';
import CardContainer from './CardContainer';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import * as api from '../api'
import { mockPeople, mockVehicles, mockPlanets, cleanPeopleMock, cleanPlanetsMock, cleanVehiclesMock } from '../mockData'

describe('CardContainer Component', () => {
  beforeAll(() => {


    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({
          scroll: mockPeople
        })
      })
    })
   api.getPeople = () => api.getMorePeopleData(mockPeople)
   api.getMorePeopleData = () => (cleanPeopleMock)
   api.getPlanets = () => api.getMorePlanetData(mockPlanets)
   api.getMorePlanetData = () => (cleanPlanetsMock)
   api.getVehicles = () => (mockVehicles)
  })


  test('that it renders without crashing', () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    expect(renderedComp).toMatchSnapshot()
  })

  test('that it should update state when it gets mounted', () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    expect(renderedComp.state('people')).toEqual(cleanPeopleMock)
  })

  test('that local storage gets set upon mounting', () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    expect(localStorage.__STORE__.length).toBe(1);
  })

  test('that local storage gets set with the new props on component updating', async () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    await renderedComp.setProps({ name: 'planets', fetch: api.getPlanets });
    renderedComp.update()

    expect(localStorage.__STORE__.length).toBe(2)
  })

  test('that state updates when new props get passed down', async () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    await renderedComp.setProps({ name: 'planets', fetch: api.getPlanets });
    renderedComp.update()

    expect(renderedComp.state('planets')).toEqual(cleanPlanetsMock)
  })

  test('that the cards render accordingly when state changes/new props are received', async () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)
    expect(renderedComp).toMatchSnapshot()

    await renderedComp.setProps({ name: 'planets', fetch: api.getPlanets });
    renderedComp.update()
    expect(renderedComp).toMatchSnapshot()

  })

  test('that if new props are passed twice, local storage should have 3 key/value pairs', async () => {
    const renderedComp = shallow(<CardContainer name='people' fetch={api.getPeople} />)

    await renderedComp.setProps({ name: 'planets', fetch: api.getPlanets });
    renderedComp.update()

    await renderedComp.setProps({ name: 'vehicles', fetch: api.getVehicles });
    renderedComp.update()

    expect(localStorage.__STORE__.length).toBe(3)
  })

})