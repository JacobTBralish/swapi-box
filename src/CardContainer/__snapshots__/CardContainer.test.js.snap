// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer Component that it renders without crashing 1`] = `
<div
  className="card-container"
>
  <h1>
     Loading... 
  </h1>
</div>
`;

exports[`CardContainer Component that the cards render accordingly when state changes/new props are received 1`] = `
<div
  className="card-container"
>
  <Component
    Name="Luke Skywalker"
    handleClick={[Function]}
    id="Luke Skywalker"
    key="Luke Skywalker0"
  />
</div>
`;

exports[`CardContainer Component that the cards render accordingly when state changes/new props are received 2`] = `
<div
  className="card-container"
>
  <Component
    Name="Tatooine"
    Population="200000"
    Residents={
      Array [
        "undefined ",
        "undefined ",
      ]
    }
    Terrain="desert"
    handleClick={[Function]}
    id="Tatooine"
    key="Tatooine0"
  />
</div>
`;
